[atable] Feat sorting and filtering

lauty95 opened the pull request:
# Sorting and Filtering

https://github.com/agritheory/stonecrop/issues/258

## Demo

The demo video shows the implementation using example code in `examples/atable/default.story.vue`

https://github.com/user-attachments/assets/1c09e7b5-645c-4c68-835f-d3ee521a5c8c

Includes styling contributions from @crabinak in https://github.com/agritheory/stonecrop/issues/258#issuecomment-3344955873

## Changes

I renamed the ATableFilter.vue component to ATableColumnFilter.vue. This new component renders the individual column filters directly beneath their corresponding header titles in ATableHeader.vue.

Additionally, the column title in ATableHeader.vue now includes the handleSort function, which works when the column is sortable.

lauty95 committed:
feat: sorting

lauty95 committed:
feat: filters component

lauty95 committed:
feat: using ATableFilters component

lauty95 committed:
feat: filter function

lauty95 committed:
feat: using ATableFilters component

lauty95 committed:
feat: Variant to story

lauty95 committed:
md

lauty95 committed:
changes

lauty95 committed:
Merge branch 'development' into feat-sorting_and_filtering

lauty95 committed:
md

lauty95 committed:
feat: ATableHeader to ATableColumnFilter

Alchez commented on common/changes/@stonecrop/atable/feat-sorting_and_filtering_2025-09-11-13-00.json:
```diff
@@ -0,0 +1,10 @@
+    {
+      "packageName": "@stonecrop/atable",
+      "comment": "Sorting and Filtering feature added by props",
+      "type": "none"
```
Let's change to "patch"


```suggestion
      "type": "patch"
```

Alchez commented on atable/src/components/ATableColumnFilter.vue:
```diff
@@ -0,0 +1,212 @@
+const filterStartValue = ref<any>('')
+const filterEndValue = ref<any>('')
+
+const getFilterType = (column: TableColumn): string => column.filterType || 'text'
```
Let's make this a `computed` property since it only needs to be resolved once per component trigger

lauty95 replied:
I reviewed it. I'm not sure it's necessary to make that function a computed property since it's a simple function.
I removed it and now I'm just using `column.filterType` directly in the v-if instead of calling `getFilterType()`.

What do you think?

Alchez replied:
Yeah, that's even better 👍

Alchez commented on atable/src/components/ATableColumnFilter.vue:
```diff
@@ -0,0 +1,212 @@
+		filterValue.value = ''
+	} else {
+		filterValue.value = value
+		store.setFilter(colIndex, { type, value })
```
I'm not sure where the `type` is being used for the filter, is this required?

lauty95 replied:
Good question.
At the beginning I thought it might be necessary, but after reviewing it I agree that the type prop isn't needed, so I removed it.

Alchez commented on atable/src/components/ATableColumnFilter.vue:
```diff
@@ -0,0 +1,212 @@
+
+const filterValue = ref<any>('')
+const filterStartValue = ref<any>('')
+const filterEndValue = ref<any>('')
```
Since these are related, I wonder if it's better to use a `reactive` instead?


```suggestion
const dateFilter = reactive({
	startValue: '' as string,
	endValue: '' as string,
})
```

lauty95 replied:
Applied!

Alchez reviewed (changes requested):
@lauty95 This looks great, but I think we should look into the following:

- [UX] The filter input boxes should maybe have a placeholder value of the column you're filtering (ex. Homepage)
- [UX] The date range filtering shouldn't start until the user has selected a date, because currently switching views to a different month/year automatically picks up a date and applies the filter
- [UI] I think the theming should maybe adjust for contrast, since I'm unable to see the clear filters/clear sort buttons on white background pages.

<img width="1330" height="185" alt="Image" src="https://github.com/user-attachments/assets/2f0bd4e8-69a1-40d6-8d21-d3ab442f525f" />

PS: If possible, let's also add tests for these additions. And check the failing tests for verify changes.

CC: @agritheory

lauty95 committed:
change type patch

Co-authored-by: Rohan <Alchez@users.noreply.github.com>

lauty95 committed:
feat: opts examples

lauty95 committed:
fix: dateFilter reactive

lauty95 committed:
fix: getFilterType deleted

lauty95 committed:
fix: type parameter deleted

lauty95 committed:
changes

lauty95 committed:
fix: change input event to change event on dates

lauty95 committed:
fix: styles clear btn

lauty95 commented:
> [UX] The filter input boxes should maybe have a placeholder value of the column you're filtering (ex. Homepage)

Each filter is displayed under its corresponding header, and I also based the design on Curtis's reference: https://github.com/agritheory/stonecrop/issues/258#issuecomment-3344955873

Are you sure we should include a placeholder?

> [UX] The date range filtering shouldn't start until the user has selected a date, because currently switching views to a different month/year automatically picks up a date and applies the filter

Absolutely agree. I changed the @input event on `<input type="date" />` to @change, which allows navigating between months or years without triggering the filter until a date is actually selected.

> [UI] I think the theming should maybe adjust for contrast, since I'm unable to see the clear filters/clear sort buttons on white background pages.

What about something like this:

<img width="1317" height="167" alt="Screenshot 2025-11-25 at 11 05 36 AM" src="https://github.com/user-attachments/assets/5b6f1477-cc89-4fd1-b83d-f105b9f5cbdc" />

<img width="1323" height="194" alt="Screenshot 2025-11-25 at 11 06 04 AM" src="https://github.com/user-attachments/assets/4bfec523-0db0-49c4-bd5c-fba664b2055a" />

Alchez commented:
@lauty95 

> > [UX] The filter input boxes should maybe have a placeholder value of the column you're filtering (ex. Homepage)
> 
> Each filter is displayed under its corresponding header, and I also based the design on Curtis's reference: [#258 (comment)](https://github.com/agritheory/stonecrop/issues/258#issuecomment-3344955873)
> 
> Are you sure we should include a placeholder?

That's a good point too. I think it's fine if we skip a placeholder for now.

> > [UI] I think the theming should maybe adjust for contrast, since I'm unable to see the clear filters/clear sort buttons on white background pages.
> 
> What about something like this:
> <img alt="Screenshot 2025-11-25 at 11 05 36 AM" width="1317" height="167" src="https://private-user-images.githubusercontent.com/38353324/518677326-5b6f1477-cc89-4fd1-b83d-f105b9f5cbdc.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQwODI3NzQsIm5iZiI6MTc2NDA4MjQ3NCwicGF0aCI6Ii8zODM1MzMyNC81MTg2NzczMjYtNWI2ZjE0NzctY2M4OS00ZmQxLWI4M2QtZjEwNWI5ZjVjYmRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTI1VDE0NTQzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlM2JmNDY4MGFhZGQ4ZWMzNGJmZjhmYjdmYzgzNzAwMTU2NDhmZTIyZmVmZWMxOGQyYjNhOTYxMWVmZjkyY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Sexbu1jW0PvFvK3zXsaJ9aWSLSJ0cxhCxVsec3YNKMY"> <img alt="Screenshot 2025-11-25 at 11 06 04 AM" width="1323" height="194" src="https://private-user-images.githubusercontent.com/38353324/518677566-4bfec523-0db0-49c4-bd5c-fba664b2055a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQwODI3NzQsIm5iZiI6MTc2NDA4MjQ3NCwicGF0aCI6Ii8zODM1MzMyNC81MTg2Nzc1NjYtNGJmZWM1MjMtMGRiMC00OWM0LWJkNWMtZmJhNjY0YjIwNTVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTI1VDE0NTQzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzYzgxYmI2YjliMDFkYTEwMzFiMTE0MDZjMDZkMTU4YTdmZTNiNTBmNDM1N2ZiYTFjYzkyZjNkMjhiZTE2NWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6PFQ_3-xV8jhNuAsY0eTYFSUDb9SigpoSYtIjWXSvjQ">

Yeah, that looks great!

Alchez commented on atable/src/stores/table.ts:
```diff
@@ -186,6 +186,20 @@ export const createTableStore = (initData: {
+					endValue?: any
+				}
+			>
+		>({})
```
This seems very hard to read, can we extract this into a separate type instead?

lauty95 replied:
ready!

Alchez commented on common/changes/@stonecrop/aform/feat-sorting_and_filtering_2025-11-25-13-12.json:
```diff
@@ -0,0 +1,10 @@
+    {
+      "packageName": "@stonecrop/aform",
+      "comment": "Edited API.md to document new sorting and filtering features in ATable component.",
+      "type": "none"
```
"patch" on these ones too

Alchez reviewed (changes requested):
I think this is pretty close. Let's also add tests, since our coverage has dropped dramatically 😅 

PS: I'm headed on vacation until Dec 2, so I won't be able to test this out by myself, but I could probably try give you more feedback via mobile.

lauty95 commented:
Thanks for your review, @Alchez! I'll add some tests for these features.

Enjoy your vacation!

lauty95 committed:
fix: type patch

lauty95 committed:
test: Column Filter Component

lauty95 committed:
test: sort and filter functionality

lauty95 committed:
test: AForm coverage

lauty95 committed:
refactor: filter state definition

lauty95 commented:
I added some tests to cover sorting and filtering

Also added some tests in `aform/tests/dropdown.spec.ts` and `aform/tests/form.spec.ts` to raise the coverage from 65% to 70% and avoid the following error:

```sh
Coverage for branches (65.46%) does not meet global threshold (70%)
```

Alchez committed:
chore: fix lint issues

Alchez committed:
fix: set next version bump type
